# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [10]:
import pandas as pd
import numpy as np
from seaborn import load_dataset

import patsy
import statsmodels.api as sm
import statsmodels.formula.api as smf

import seaborn as sns
import matplotlib.pyplot as plt


In [11]:
tips = sns.load_dataset("tips")
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']
tips.head()

total_bill   tip     sex smoker  day    time  size   tip_pct  net_bill
0       16.99  1.01  Female     No  Sun  Dinner     2  0.063204     15.98
1       10.34  1.66    Male     No  Sun  Dinner     3  0.191244      8.68
2       21.01  3.50    Male     No  Sun  Dinner     3  0.199886     17.51
3       23.68  3.31    Male     No  Sun  Dinner     2  0.162494     20.37
4       24.59  3.61  Female     No  Sun  Dinner     4  0.172069     20.98

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [5]:
# Carregando a base de dados 'tips'
tips = sns.load_dataset("tips")

# Criando as variáveis 'tip_pct' e 'net_bill'
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']

# Adicionando o logaritmo de net_bill
tips['log_net_bill'] = np.log(tips['net_bill'])

# Ajustando o modelo de regressão para tip_pct em função do logaritmo de net_bill
model_log = smf.ols('tip_pct ~ log_net_bill', data=tips).fit()

# Exibindo os resultados do modelo
print(model_log.summary())


                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.198
Method:                 Least Squares   F-statistic:                     61.16
Date:                Fri, 24 Nov 2023   Prob (F-statistic):           1.62e-13
Time:                        09:49:41   Log-Likelihood:                 123.81
No. Observations:                 244   AIC:                            -243.6
Df Residuals:                     242   BIC:                            -236.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        0.6181      0.054     11.444   

 3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [6]:
import pandas as pd
import numpy as np
import seaborn as sns
import statsmodels.formula.api as smf

# Carregando a base de dados 'tips'
tips = sns.load_dataset("tips")

# Criando as variáveis 'tip_pct' e 'net_bill'
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']

# Ajustando o modelo de regressão de tip_pct em função de um polinômio de segundo grau em net_bill
model_poly = smf.ols('tip_pct ~ I(net_bill) + I(net_bill ** 2)', data=tips).fit()

# Exibindo o resumo do modelo
print(model_poly.summary())


                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.158
Model:                            OLS   Adj. R-squared:                  0.151
Method:                 Least Squares   F-statistic:                     22.55
Date:                Fri, 24 Nov 2023   Prob (F-statistic):           1.06e-09
Time:                        09:55:13   Log-Likelihood:                 117.25
No. Observations:                 244   AIC:                            -228.5
Df Residuals:                     241   BIC:                            -218.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.4851      0.047  

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [8]:
# Carregando a base de dados 'tips'
tips = sns.load_dataset("tips")

# Criando as variáveis 'tip_pct' e 'net_bill'
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']

# Verificando se há valores de tip_pct menores ou iguais a zero
if (tips['tip_pct'] <= 0).any():
    print("Existem valores de tip_pct menores ou iguais a zero, o que impede o cálculo do logaritmo.")
else:
    # Calculando o logaritmo de tip_pct
    tips['log_tip_pct'] = np.log(tips['tip_pct'])

    # Ajustando o modelo de regressão para o logaritmo de tip_pct em função de net_bill
    model_log_tip_pct = smf.ols('log_tip_pct ~ net_bill', data=tips).fit()

    # Calculando o R^2 para tip_pct usando os valores previstos e convertendo de volta do logaritmo
    tips['pred_log_tip_pct'] = model_log_tip_pct.fittedvalues
    tips['pred_tip_pct'] = np.exp(tips['pred_log_tip_pct'])

    # Calculando o R^2 para tip_pct
    ss_res = np.sum(np.square(tips['tip_pct'] - tips['pred_tip_pct']))
    ss_tot = np.sum(np.square(tips['tip_pct'] - np.mean(tips['tip_pct'])))
    r_squared = 1 - (ss_res / ss_tot)

    print(f"O R^2 para o modelo no log de tip_pct é: {model_log_tip_pct.rsquared}")
    print(f"O R^2 calculado para tip_pct é: {r_squared}")


O R^2 para o modelo no log de tip_pct é: 0.23534487140864324
O R^2 calculado para tip_pct é: 0.08676609740629249


###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [9]:
# Carregar a base de dados 'tips'
tips = sns.load_dataset("tips")

# Criar as variáveis 'tip_pct' e 'net_bill'
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']

# Remover valores de tip_pct e net_bill menores ou iguais a zero
tips = tips[(tips['tip_pct'] > 0) & (tips['net_bill'] > 0)]

# Calcular o logaritmo de tip_pct e net_bill
tips['log_tip_pct'] = np.log(tips['tip_pct'])
tips['log_net_bill'] = np.log(tips['net_bill'])

# Ajustar o modelo de regressão linear
model = smf.ols('log_tip_pct ~ log_net_bill', data=tips).fit()

# Calcular o R^2 do modelo
r_squared_model = model.rsquared

# Prever o log de tip_pct e converter para a escala original
tips['pred_log_tip_pct'] = model.predict(tips[['log_net_bill']])
tips['pred_tip_pct'] = np.exp(tips['pred_log_tip_pct'])

# Calcular o R^2 para tip_pct
ss_res = np.sum(np.square(tips['tip_pct'] - tips['pred_tip_pct']))
ss_tot = np.sum(np.square(tips['tip_pct'] - np.mean(tips['tip_pct'])))
r_squared_tip_pct = 1 - (ss_res / ss_tot)

print(f"R^2 do modelo no log de tip_pct: {r_squared_model}")
print(f"R^2 calculado para tip_pct: {r_squared_tip_pct}")


R^2 do modelo no log de tip_pct: 0.29347402150807766
R^2 calculado para tip_pct: 0.22768529518136027


###  6. qual modelo resultou no melhor ajuste?

O modelo, com um R2 de 0.293 para o log de tip_pct e 0.228 para tip_pct na escala original, apresenta um ajuste melhor aos dados do que o outro modelo, que tem um R2 de 0.235 para o log de tip_pct e 0.087 para tip_pct na escala original.